Start of e2e testing support for Wasp (Part 1)

shayneczyzewski opened the pull request:
# Description

Start of golden testing approach for end-to-end testing for Wasp. Uses this library: https://hackage.haskell.org/package/tasty-golden to compare Golden Directories vs some test directories.

This PR adds a golden testing capability for the following commands:
- wasp build
- wasp compile
- wasp migrate
- wasp new

Note: I currently am not storing or checking stdout, nor wasp start, in any way (that can be future work I think)

**There is a second part of this work which is adding the golden directory output, covered here: PR #475** 

## Type of change

Please select the option(s) that is more relevant.

- [ ] Code cleanup
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

matijaSos commented on waspc/e2e-test/Util.hs:
```diff
@@ -0,0 +1,82 @@
+runGoldenTest :: FilePath -> String -> (String -> String) -> IO TestTree
+runGoldenTest cwd goldenDirName command = do
+  let testDirName = goldenDirName ++ "Test"
+  let command' = command testDirName
```
this is later used only once, so maybe not worth extracting it?

shayneczyzewski replied:
Good point, I could have also renamed the identifier to be `commandRunningRelativeToTestOutputDir` lol ok maybe not, but that is what it does when you give it the dir, constructs all commands relative to that FP. I'll just not extract for now.

matijaSos replied:
makes sense! although I like your long name also :D

matijaSos reviewed (commented):
woohoo great job, really enjoyed reading the code! Left the initial round of comments. Really excited about this feature.

Martinsos commented on waspc/e2e-test/Util.hs:
```diff
@@ -0,0 +1,82 @@
+          (\ref new -> ["diff", "-u", ref, new])
+          goldenFilePath
+          generatedFp
+          (return ()) -- this is fine :D
```
I like the :D and it made me smile but I am also not sure why is it fine :D.

shayneczyzewski replied:
Haha will detail in comment, wasn't sure of the review level for the draft but had lots of this kinda cruft remaining. Cleaning up now!

Martinsos replied:
![image](https://user-images.githubusercontent.com/1536647/154939889-5317ea48-76b7-4e3b-8f7d-7a9161083970.png)

Martinsos commented on waspc/e2e-test/Util.hs:
```diff
@@ -0,0 +1,82 @@
+      return $ notElem (takeFileName fp) [".waspinfo", ".gitignore", "node_modules", "dev.db", "dev.db-journal", "package.json", "package-lock.json", "golden.manifest"]
+
+writeManifest :: FilePath -> [FilePath] -> FilePath -> IO FilePath
+writeManifest cwd filePaths testDirAbsFp = do
```
+1 on this

shayneczyzewski replied:
Good point. I tried to clean up by letting caller give us the file paths to write, and the place to write it, so it has ultimate say (and can ensure consistency). It then is now a local where helper that does:

```haskell
writeFileManifest relToAbsFp filePaths manifestAbsFp = do
  let sortedRelativeFilePaths = unlines . sort . map (makeRelative relToAbsFp) $ filePaths
  writeFile manifestAbsFp sortedRelativeFilePaths
```

Martinsos commented on waspc/e2e-test/Util.hs:
```diff
@@ -0,0 +1,82 @@
+writeManifest :: FilePath -> [FilePath] -> FilePath -> IO FilePath
+writeManifest cwd filePaths testDirAbsFp = do
+  let sortedRelativeFilePaths = unlines . sort . map (makeRelative cwd) $ filePaths
+  let manifestFp = testDirAbsFp </> "golden.manifest"
```
Or `testOutput.manifest`.

shayneczyzewski replied:
Good point. I kinda like `files.manifest` 👍🏻

Martinsos commented on waspc/e2e-test/Util.hs:
```diff
@@ -0,0 +1,82 @@
+import Data.List (intercalate, sort)
+import Data.Text (pack, replace, unpack)
+import System.Directory (doesFileExist)
+import System.Directory.Recursive (getDirFiltered)
```
What about listDirectoryDeep that we have in Wasp.Util.IO -> would that maybe be a better choice than using `getDirFiltered`? I am not sure, just mentioning it in case you are not aware that we have that function.

shayneczyzewski replied:
Yeah, I remembered that one. I had not moved to SP yet so `getDirFiltered` was nice. It also supported filters, which `listDirectoryDeep` does not appear to. I could filter after, but going into `node_modules` unnecessarily could be an expensive proposition.

Martinsos replied:
Ah uff, ok that is very important indeed -> if `getDirFiltered` does filtering while searching, which it does, that is very important due to `node_modules` as you mentioned. Makes sense then, certainly stick with it!

Martinsos reviewed (commented):
@shayneczyzewski this is super cool all together :)!!

Left some comments, lot of small stuff, nothing major, all makes sense in general.

One question I wasn't sure about: where is test output generated? Is it under `golden` directory? If so, why not separate directory, would that make stuff easier?

shayneczyzewski commented:
Thanks for the review and thoughts on this early draft @Martinsos and @matijaSos! I will work on replying/addressing the comments as I finish out the remaining functionality and add a few more golden test scenarios over the next few days. 🚀

shayneczyzewski commented on waspc/e2e-test/Main.hs:
```diff
@@ -0,0 +1,43 @@
+  t3 <- waspMigrateGoldenTest cwd
+  return $ testGroup "All Golden Dir Tests" [t1, t2, t3]
+
+waspNewGoldenTest :: FilePath -> IO TestTree
```
Indeed they all do currently have that shape. I like the idea of a type now and then newType later (potentially making it a ReaderT and tossing cwd into it), and redefining both the names and types of the functions. Good idea!

shayneczyzewski commented on waspc/e2e-test/Util.hs:
```diff
@@ -0,0 +1,82 @@
+
+-- TODO: Handle shell command descriptions/invocations and file manipulation in some better way (and more platform agnostic?)
+-- NOTE: Using `wasp-cli` herein so we can assume using latest `stack install` in CI and locally. Better way?
+newCommand :: String -> String
```
Good ideas here, will incorporate. 👍🏻 Overall I'd like to make the chain-able in some way so you can compose them together more easily, but baby steps. :)

shayneczyzewski commented on waspc/e2e-test/Util.hs:
```diff
@@ -0,0 +1,82 @@
+-- NOTE: Using `wasp-cli` herein so we can assume using latest `stack install` in CI and locally. Better way?
+newCommand :: String -> String
+newCommand projectName =
+  intercalate " && " ["wasp-cli new " ++ projectName, "cp " ++ ("overlays" </> ".gitignore") ++ " " ++ projectName]
```
Yep sounds good!

shayneczyzewski commented on waspc/e2e-test/Util.hs:
```diff
@@ -0,0 +1,82 @@
+  intercalate " && " [newCommand projectName, "cd " ++ projectName, "wasp-cli compile"]
+
+newAndCompileAndMigrateCommand :: FilePath -> FilePath -> String -> String
+newAndCompileAndMigrateCommand _cwd overlayFp projectName =
```
Yeah I kinda made that up, since to me they are not fixtures in the traditional sense how I'm (abusing? lol) using them... I am using them sometimes to overwrite an entire file (like overlays/.gitignore), sometimes to append content to another file (like overlays/entity.txt), so the idea of "overlaying files/contents onto existing ones" was kinda where I was going. This whole area can probably be cleaned up some.

Martinsos replied:
Can you maybe help me understand better what these are used for? What are the use cases you use them for, in more details (ELI5).

shayneczyzewski replied:
The thinking here was having an easy way to add discrete things to Wasp files, like entities. This is in contrast to having a full Wasp file we overwrite. I thought the applying patches approach was a bit more flexible and more maintainable.

Martinsos replied:
Ah got it! Ok, so we can ensure that for example Wasp file has an entity? Makes sense! Might be interesting to find a better name, if we can.

shayneczyzewski replied:
Yeah agreed,  not sure what the convention here is, will think on it and at least better comment the purpose in meantime 🤔 👍🏻

shayneczyzewski commented on waspc/e2e-test/Util.hs:
```diff
@@ -0,0 +1,82 @@
+   in intercalate
+        " && "
+        [ newAndCompileCommand projectName,
+          "cat " ++ overlayFp ++ " >> main.wasp",
```
Yep more comments here will help. For that specific one: `-- Append content from the overlay file (e.g. an entity) to the end of the wasp file`

Martinsos replied:
Aha! And why are you doing that?

shayneczyzewski commented on waspc/e2e-test/Util.hs:
```diff
@@ -0,0 +1,82 @@
+runGoldenTest :: FilePath -> String -> (String -> String) -> IO TestTree
+runGoldenTest cwd goldenDirName command = do
+  let testDirName = goldenDirName ++ "Test"
+  let command' = command testDirName
```
That is so we can construct all commands relative to some test directory (like create a new project of that name, cd into it, do other ops relative to it).

shayneczyzewski replied:
But yes, it is actually both, the project name, and the resulting directory from which other things need a stable reference.

shayneczyzewski commented on waspc/e2e-test/Util.hs:
```diff
@@ -0,0 +1,82 @@
+
+  -- (1) Run Wasp CLI and other command(s)
+  -- TODO: Save stdout/error as log file for "contains" checks
+  let testDirAbsFp = cwd </> "e2e-test" </> "goldens" </> testDirName
```
Generated project dir, I can try to clean that up.

Martinsos replied:
Sounds good, I think using better names will already help a lot.

shayneczyzewski commented on waspc/e2e-test/Util.hs:
```diff
@@ -0,0 +1,82 @@
+  -- TODO: Save stdout/error as log file for "contains" checks
+  let testDirAbsFp = cwd </> "e2e-test" </> "goldens" </> testDirName
+
+  callCommand $ "rm -rf " ++ testDirAbsFp
```
The test dir in this case is really just the wasp project we create and manipulate.

Martinsos replied:
So why do we delete it first? To makes sure there are no left-overs from previous runs?

shayneczyzewski replied:
Correct, each run will use the same path to some test directory for that run vs a golden test. I can add a comment.

shayneczyzewski commented on waspc/e2e-test/Util.hs:
```diff
@@ -0,0 +1,82 @@
+  let testDirAbsFp = cwd </> "e2e-test" </> "goldens" </> testDirName
+
+  callCommand $ "rm -rf " ++ testDirAbsFp
+  callCommand $ "cd " ++ (cwd </> "e2e-test" </> "goldens") ++ " && " ++ command'
```
Good idea on more comments here.

shayneczyzewski commented on waspc/e2e-test/Util.hs:
```diff
@@ -0,0 +1,82 @@
+  callCommand $ "cd " ++ (cwd </> "e2e-test" </> "goldens") ++ " && " ++ command'
+
+  -- (2) Find all files in the generated test dir & store manifest to detect file additions/deletions
+  generatedFps <- getDirFiltered dirFilter testDirAbsFp >>= filterM doesFileExist
```
Yeah good idea on helper 👍🏻 Comments were stale from pairing. Had not done a full cleanup or review yet for the draft.

Martinsos replied:
Ok sorry in that case for rushing with the comments! I was excited to see how golden tests work :D.

shayneczyzewski replied:
Haha no sweat!! I love the excitement here :D

shayneczyzewski commented on waspc/e2e-test/Util.hs:
```diff
@@ -0,0 +1,82 @@
+    testGroup
+      goldenDirName
+      [ goldenVsFileDiff
+          generatedFp -- test name
```
True, it is messy- I will try to clean these up some

Martinsos replied:
"output" vs "golden" is not so bad I think.

If we really wanted to be verbose, it would be "currentOutput" vs "goldenOutput".

shayneczyzewski replied:
I was moving toward "testProject" vs "golden", but don't have a strong preference to that direction atm.

shayneczyzewski commented on waspc/e2e-test/Util.hs:
```diff
@@ -0,0 +1,82 @@
+          goldenFilePath
+          generatedFp
+          (return ()) -- this is fine :D
+        | generatedFp <- manifestFp : generatedFps,
```
Manifest gets generated from all files in the test project and then compared against the golden one, yep. This detects added/removed files.

shayneczyzewski commented on waspc/e2e-test/Util.hs:
```diff
@@ -0,0 +1,82 @@
+          generatedFp
+          (return ()) -- this is fine :D
+        | generatedFp <- manifestFp : generatedFps,
+          let goldenFilePath = unpack $ replace (pack testDirName) (pack goldenDirName) (pack generatedFp) -- (3) TODO: remap test dir to golden dir paths in some better way
```
String patches for paths just feels icky, but with SP I will clean this up 👍🏻

shayneczyzewski commented on waspc/e2e-test/Util.hs:
```diff
@@ -0,0 +1,82 @@
+      return $ notElem (takeFileName fp) [".waspinfo", ".gitignore", "node_modules", "dev.db", "dev.db-journal", "package.json", "package-lock.json", "golden.manifest"]
+
+writeManifest :: FilePath -> [FilePath] -> FilePath -> IO FilePath
+writeManifest cwd filePaths testDirAbsFp = do
```
yeah, they are `testProjectFps`, still trying to get all the naming straight 😅

shayneczyzewski reviewed (commented):
Started to implement some of the feedback. To answer your other comment question @Martinsos about where the test projects live: they do indeed end up under golden (are gitignored) and mainly there for easy diffing.

The folder structure is like:
```
e2e-test/goldens
├── overlays
├── waspCompile
├── waspCompileTest
├── waspMigrate
├── waspMigrateTest
├── waspNew
└── waspNewTest
```

The rationale for that was to make the diffing easier, so you can just go into goldens and do: `<difftool> waspCompile*` to resolve the golden vs test project issues for the `wasp compile` test. However, open to ideas on where they should live to make it easier to resolve 👍🏻 thnx all

Martinsos commented on waspc/e2e-test/Main.hs:
```diff
@@ -0,0 +1,43 @@
+main :: IO ()
+main = do
+  -- Skip Windows in CI for now
+  unless (os == "mingw32") $ do
```
Ah I see what you mean! Makes sense!

So on one hand we have e2e tests being silently skipped on Windows, which doesn't sound great, but on the other hand, we have `stack test` failing on Windows, which is also not fun.
At the end both cases are not really important because I doubt much/any development will be done directly on Windows.
Ok, whatever you choose to do is fine with me!

shayneczyzewski replied:
For now I put it into a message to the user that should hopefully inform the intent without a comment:

```haskell
if os == "mingw32"
    then putStrLn "Skipping end-to-end tests on Windows due to tests using *nix-only commands"
```

I'l chew on if a comment about erroring out vs not (and `stack test` implications) makes sense to add 👍🏻

Martinsos replied:
I think this is already great!

Martinsos commented:
> Started to implement some of the feedback. To answer your other comment question @Martinsos about where the test projects live: they do indeed end up under golden (are gitignored) and mainly there for easy diffing.
> 
> The folder structure is like:
> 
> ```
> e2e-test/goldens
> ├── overlays
> ├── waspCompile
> ├── waspCompileTest
> ├── waspMigrate
> ├── waspMigrateTest
> ├── waspNew
> └── waspNewTest
> ```
> 
> The rationale for that was to make the diffing easier, so you can just go into goldens and do: `<difftool> waspCompile*` to resolve the golden vs test project issues for the `wasp compile` test. However, open to ideas on where they should live to make it easier to resolve 👍🏻 thnx all

Tahnks for explanation, now I have a better idea!

I thought it might be nice to have a way to be able to differentiate golden test output from current test output. Maybe by putting golden ones into `golden`, so you have `golden/waspNew`, `golden/WaspMigrate`, while current test output would be in `current`, so you would have `current/waspNew` and `current/waspMigrate`.

That said, I haven't paid attention to ease of diffing! If you think this makes diffing harder than it needs to be, that is a good argument. In that case, alternative might also be to have them all in the same directory as is now, but name them like `waspNew_golden`, `waspMigrate_golden`, `waspNew_current`, `waspMigrate_current`?

faassen commented on waspc/cli/Wasp/Cli/Command/Db/Migrate.hs:
```diff
@@ -21,8 +21,8 @@ import qualified Wasp.Generator.DbGenerator.Operations as DbOps
 -- The migrate function takes care of copying migrations from the generated project back to the source code.
-migrateDev :: Command ()
-migrateDev = do
+migrateDev :: Maybe String -> Command ()
```
How does this change relate to end to end testing?

faassen commented on waspc/e2e-test/Main.hs:
```diff
@@ -0,0 +1,54 @@
+      -- Stack/Cabal launch `stack test` from root of project, so this should always be some absolute path to waspc
+      cwd <- getCurrentDirectory
+      unless (takeFileName cwd == "waspc") (error "Expecting test process to be invoked from waspc dir")
+      defaultMain =<< tests cwd
```
Since this is already in a do block, can't you just do:

```
tests cwd
defaultMain
```
?

Martinsos replied:
Hijacking the answer, but actually the equivalent would be
```hs
x <- tests cwd
defaultMain x
```

`=<<` forward the result of the action on the right to the action on the left.

If it was `<<` instead, then what you proposed would be valid replacement, since result of `tests` would be discarded.

faassen commented on waspc/e2e-test/Main.hs:
```diff
@@ -0,0 +1,54 @@
+
+waspNew :: GoldenTest
+waspNew cwd =
+  runGoldenTest cwd "waspNew" waspCliNew
```
Extracting functions into another module hides what the test actually does from the place where the test gets run. I'm a bit concerned about that.

Given the amount of repetition left, I'd be inclined to investigate whether we can declare tests as a list:

```haskell
[("waspNew", waspCliNew), ("waspCompile", waspCliNewAndCompile)]
```

and have a test runner.

shayneczyzewski replied:
I think we were able to chat this one through on the call

faassen commented on waspc/e2e-test/Main.hs:
```diff
@@ -0,0 +1,54 @@
+
+waspMigrate :: GoldenTest
+waspMigrate cwd =
+  runGoldenTest cwd "waspMigrate" (waspCliNewAndCompileAndMigrate cwd (cwd </> joinPath ["e2e-test", "goldens", "overlays", "entity.txt"]))
```
Here `waspCliNewAndCompileAndMigrate` takes a lot more arguments, and I'm not sure why. I think if the test was inlined and flatter (not a separate function in another file) it'd be easier for me to understand what's going on.

(with tests in my opinion many rules of factoring change; simple flat tests often are easier to understand and maintain than test code split into functions)

shayneczyzewski replied:
That is the overlay file, but yeah arguably putting too much logic into the commands and less into the tests is not ideal. The pro to extracting is the commands can be used in other, larger tests. We can chat more on it?

faassen commented on waspc/e2e-test/Util.hs:
```diff
@@ -0,0 +1,82 @@
+module Util
```
I think the tests should go back into `Main.hs` (or whatever the name should be; I'd have expected something with the word "test" in it).

faassen replied:
Instead of "Util" we can call the rest "GoldenTest.hs" or something. Should that be in this directory?

shayneczyzewski replied:
Came up with the following by renaming `Main.hs` with `E2eTestDriver.hs`, splitting `Util.hs` into `GoldenTest.hs` and `ShellCommands.hs`, and putting tests into their own (tiny lol) files:

```bash
e2e-test/
├── E2eTestDriver.hs
├── GoldenTest.hs
├── ShellCommands.hs
├── Tests
│   ├── WaspCompileTest.hs
│   ├── WaspMigrateTest.hs
│   └── WaspNewTest.hs
└── goldens
    ├── overlays
    ├── waspCompile
    ├── waspCompileTest
    ├── waspMigrate
    ├── waspMigrateTest
    ├── waspNew
    └── waspNewTest
```

We can iterate perhaps more on a quick sync call to resolve any concerns about this layout.

shayneczyzewski replied:
The one thing that would be nice is if I could use Tasty discover to find the actual tests, but I wasn't sure how to do so given what I need to do in my `main`, but may try again after a bit.

faassen commented on waspc/cli/Wasp/Cli/Command/Db/Migrate.hs:
```diff
@@ -21,8 +21,8 @@ import qualified Wasp.Generator.DbGenerator.Operations as DbOps
-migrateDev :: Command ()
-migrateDev = do
+migrateDev :: Maybe String -> Command ()
+migrateDev maybeMigrationName = do
```
Would it be possible to require the name here and to ensure higher up that a default name is supplied if not given? Maybe that's tricky to accomplish?

shayneczyzewski replied:
I'm not sure we want to use a default name. Either they gave one to us, or we let Prisma ask them interactively (current default behavior).

Martinsos replied:
I agree with this for now, to not use a default name.

matijaSos replied:
yeah, as @shayneczyzewski said, it seems that Prisma's way of working is not having default names, as it kinda encourages bad practices (losing info, I guess). So I think it will not be strange if we follow that.

faassen commented on waspc/e2e-test/Util.hs:
```diff
@@ -0,0 +1,96 @@
+          let goldenFp = unpack $ replace (pack testProjectName) (pack goldenDirName) (pack testProjectFp) -- TODO: clean up after SP transition
+      ]
+  where
+    -- TODO: Ideally we would not ignore `package.json`, but in CI the order of dependencies differs :/
```
Interesting! We should sort them before generation then. Can you create an issue?

shayneczyzewski replied:
This should be the Node package.json output from `npm install`, not your json file... can we control how Node outputs? Or do you mean we should be somehow parsing and rewriting that file during your stuff?

faassen commented on waspc/src/Wasp/Generator/DbGenerator/Jobs.hs:
```diff
@@ -26,11 +26,15 @@ import Wasp.Generator.ServerGenerator.Common (serverRootDirInProjectRootDir)
   let serverDir = projectDir </> serverRootDirInProjectRootDir
   let schemaFile = projectDir </> dbSchemaFileInProjectRootDir
 
+  let migrationArgs = case maybeMigrationName of
```
Would it be possible to supply a migrationName always on this level or is it really the responsibility lower down to generate a name if it's missing?

shayneczyzewski replied:
Migration name would come from user (via CLI invocation). If we have one we can use with the flag, but if not we would not generate one for them automatically- Prisma would ask them for a migration name interactively if they do not supply one ahead of time.

shayneczyzewski replied:
Thanks for the good tip Matija! I like yours too, Martin, but that may be too golfy for me lol :)

shayneczyzewski commented:
> > Started to implement some of the feedback. To answer your other comment question @Martinsos about where the test projects live: they do indeed end up under golden (are gitignored) and mainly there for easy diffing.
> > The folder structure is like:
> > ```
> > e2e-test/goldens
> > ├── overlays
> > ├── waspCompile
> > ├── waspCompileTest
> > ├── waspMigrate
> > ├── waspMigrateTest
> > ├── waspNew
> > └── waspNewTest
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > The rationale for that was to make the diffing easier, so you can just go into goldens and do: `<difftool> waspCompile*` to resolve the golden vs test project issues for the `wasp compile` test. However, open to ideas on where they should live to make it easier to resolve 👍🏻 thnx all
> 
> Tahnks for explanation, now I have a better idea!
> 
> I thought it might be nice to have a way to be able to differentiate golden test output from current test output. Maybe by putting golden ones into `golden`, so you have `golden/waspNew`, `golden/WaspMigrate`, while current test output would be in `current`, so you would have `current/waspNew` and `current/waspMigrate`.
> 
> That said, I haven't paid attention to ease of diffing! If you think this makes diffing harder than it needs to be, that is a good argument. In that case, alternative might also be to have them all in the same directory as is now, but name them like `waspNew_golden`, `waspMigrate_golden`, `waspNew_current`, `waspMigrate_current`?

We can't use underscores in project names just yet :) I did have golden in the name before but got some naming preference feedback from Martijn that we may want to drop that... we can all chat on it maybe to quickly resolve sometime as I start to wrap it up, not super hard to change. We should see what is best/easiest for the diffing/resolving process most likely, but that is TBD at the moment I'd say.

shayneczyzewski commented on waspc/src/Wasp/Generator/DbGenerator/Jobs.hs:
```diff
@@ -40,7 +44,7 @@ migrateDev projectDir = do
 
   -- NOTE(martin): For this to work on Mac, filepath in the list below must be as it is now - not wrapped in any quotes.
-  let npxPrismaMigrateCmd = npxPrismaCmd ++ ["migrate", "dev", "--schema", SP.toFilePath schemaFile]
+  let npxPrismaMigrateCmd = npxPrismaCmd ++ ["migrate", "dev", "--schema", SP.toFilePath schemaFile] ++ migrationArgs
```
Yeah `additionalMigrationsArgs` sounds good, thnx. 👍🏻 

Unsure what you mean by redefine schema though?

Martinsos replied:
Ah sorry, that was badly explained: Waht I meant to say is, if they provide `--schema` argument under `additionalMigrationArgs`. They could do that and mess up our `--schema` argument. But ok, probably best not to worry about it for now.

shayneczyzewski replied:
Ah ok cool, good point to at least be aware of... it may be the case that the flag would get swallowed before making it to us as well, but something to consider if we ever need to.

Martinsos commented:
> > > Started to implement some of the feedback. To answer your other comment question @Martinsos about where the test projects live: they do indeed end up under golden (are gitignored) and mainly there for easy diffing.
> > > The folder structure is like:
> > > ```
> > > e2e-test/goldens
> > > ├── overlays
> > > ├── waspCompile
> > > ├── waspCompileTest
> > > ├── waspMigrate
> > > ├── waspMigrateTest
> > > ├── waspNew
> > > └── waspNewTest
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > The rationale for that was to make the diffing easier, so you can just go into goldens and do: `<difftool> waspCompile*` to resolve the golden vs test project issues for the `wasp compile` test. However, open to ideas on where they should live to make it easier to resolve 👍🏻 thnx all
> > 
> > 
> > Tahnks for explanation, now I have a better idea!
> > I thought it might be nice to have a way to be able to differentiate golden test output from current test output. Maybe by putting golden ones into `golden`, so you have `golden/waspNew`, `golden/WaspMigrate`, while current test output would be in `current`, so you would have `current/waspNew` and `current/waspMigrate`.
> > That said, I haven't paid attention to ease of diffing! If you think this makes diffing harder than it needs to be, that is a good argument. In that case, alternative might also be to have them all in the same directory as is now, but name them like `waspNew_golden`, `waspMigrate_golden`, `waspNew_current`, `waspMigrate_current`?
> 
> We can't use underscores in project names just yet :) I did have golden in the name before but got some naming preference feedback from Martijn that we may want to drop that... we can all chat on it maybe to quickly resolve sometime as I start to wrap it up, not super hard to change. We should see what is best/easiest for the diffing/resolving process most likely, but that is TBD at the moment I'd say.

Oh wow, why can't we use underscores, what is stopping us?

Ok makes sense, I am up for discussing it quickly if needed. I don't have a specific name I want to have, I would just like it to be clear what is what when one looks at the contents of the directory.

matijaSos commented on waspc/cli/Main.hs:
```diff
@@ -124,6 +125,7 @@ printDbUsage =
         "",
         title "EXAMPLES",
         "  wasp db migrate-dev",
+        "  wasp db migrate-dev migration-name",
```
ah yes you are right, I mixed it up! But also agree it makes sense to use more "real" name instead of migration-name, although I think the context of this being an example (which I missed) is already enough. But might be fun to provide some interesting migration name, e.g. sth with spaces if that is supported? Just to immediately educate the user on what is possible. E.g. "Added User entity".

Martinsos commented on waspc/e2e-test/Common.hs:
```diff
@@ -0,0 +1,54 @@
+
+getProjectRootPath :: IO (Path' Abs (Dir ProjectRoot))
+getProjectRootPath = do
+  -- NOTE: Stack/Cabal launch `stack test` from root of project, so this should always be some absolute path to waspc
```
```suggestion
  -- NOTE: Stack/Cabal launches `stack test` from root of the project, so this should always be some absolute path to waspc.
```

Martinsos commented on waspc/e2e-test/Common.hs:
```diff
@@ -0,0 +1,54 @@
+getProjectRootPath = do
+  -- NOTE: Stack/Cabal launch `stack test` from root of project, so this should always be some absolute path to waspc
+  absCwd <- getCurrentDirectory
+  -- Just a little extra safeguard here since we are doing destructive file ops
```
Smart!

```suggestion
  -- Just a little extra safeguard here since we are doing destructive file ops.
```

Martinsos commented on waspc/e2e-test/E2eTestDriver.hs:
```diff
@@ -0,0 +1,23 @@
+import GoldenTest (GoldenTest)
```
Nice, to have file like this!

Even if this file was named `Main`, it would be ok with me. `Driver` usually doesn't have actual tests in it, it is more of a logic for `driving` the tests. But ok.

As for your question about automatic answer -> I don't know the answer right now but I wouldn't worry about it too much, manually adding tests is not really a problem in practice, from what I observed. If it becomes we will figure it out.

shayneczyzewski replied:
Sounds good. Yeah I may go back to Main until I can automatically discover them in this file. 👍🏻

matijaSos commented on waspc/e2e-test/Main.hs:
```diff
@@ -0,0 +1,43 @@
+  t1 <- waspNewGoldenTest cwd
+  t2 <- waspCompileGoldenTest cwd
+  t3 <- waspMigrateGoldenTest cwd
+  return $ testGroup "All Golden Dir Tests" [t1, t2, t3]
```
ah nice, I also forgot about this, cool to see it in action :)

shayneczyzewski commented:
Hi everyone! I have squashed all the crazy commits from the pairing, back and forth in comments and live, and my own tests and such. I think I have landed in a place that addresses the applicable written comment feedback, as well as our group meeting. But please let me know if that is not the case.

Quick recap:
- This adds a golden testing capability for the following commands:
  - wasp build
  - wasp compile
  - wasp migrate
  - wasp new
- I currently am _not_ storing or checking stdout, nor wasp start, in any way (that can be future work I think)
- We agreed to punt on splitting out shell library stuff, wasp stuff, and test helper stuff from the current notion of `ShellCommand` until a future point (if/as needed)
- Nearly all of these changes happen under a new e2e-test dir, so pretty low risk. Here are the other changes if you wanted to do a final sanity check there:
```bash
.github/workflows/ci.yaml
waspc/cli/Main.hs
waspc/cli/Wasp/Cli/Command/Db/Migrate.hs
waspc/package.yaml
waspc/src/Wasp/Generator/DbGenerator/Jobs.hs
waspc/src/Wasp/Generator/DbGenerator/Operations.hs
waspc/stack.yaml
```
- Here is the final dir structure I landed on (but only goldens/*-golden is checked in and *-current are .gitignored):
```bash
e2e-test/
├── Common.hs
├── GoldenTest.hs
├── Main.hs
├── ShellCommands.hs
├── Tests
│   ├── WaspBuildTest.hs
│   ├── WaspCompileTest.hs
│   ├── WaspMigrateTest.hs
│   └── WaspNewTest.hs
└── goldens
    ├── waspBuild-current
    │   ├── files.manifest
    │   └── waspBuild
    ├── waspBuild-golden
    │   ├── files.manifest
    │   └── waspBuild
    ├── waspCompile-current
    │   ├── files.manifest
    │   └── waspCompile
    ├── waspCompile-golden
    │   ├── files.manifest
    │   └── waspCompile
    ├── waspMigrate-current
    │   ├── files.manifest
    │   └── waspMigrate
    ├── waspMigrate-golden
    │   ├── files.manifest
    │   └── waspMigrate
    ├── waspNew-current
    │   ├── files.manifest
    │   └── waspNew
    └── waspNew-golden
        ├── files.manifest
        └── waspNew
```
- I tried to be as consistent with our golden vs current naming as I could.
- I got rid of "overlays" and just used the entity inline (I did not need to patch .gitignore anymore after all since I am filtering out already)

Martinsos commented on waspc/e2e-test/ShellCommands.hs:
```diff
@@ -0,0 +1,83 @@
+waspCliBuild _context =
+  combineShellCommands
+    [ -- Change DB to postgres by adding string at specific line so it still parses.
+      "awk 'NR==2{print \"  db: { system: PostgreSQL },\"}1' main.wasp > main.wasp.tmp",
```
A bit scary that it does this, I wouldn't expect this based on the name.

I would consider renaming it to `setDbToPSQLAndRunWaspCliBuild`.

Or make it two separate commands.

Or maybe best, do this part directly in the definition of a golden test.

This part with setting db to PostgreSQL also looks very fragile, but ok, I understand that this was the fastest way to do it now. Maybe comment on it that it is very fragile, and assuming a lot of stuff.

Martinsos replied:
It is a bit bad that it is so hard for us to modify Wasp code.
But real solution is probably to not use `wasp new` and then patch it up but instead provide fresh wasp file just for this test. But ok, we can do this in the future.

Martinsos replied:
It is a bit bad that it is so hard for us to modify Wasp code.
But real solution is probably to not use `wasp new` and then patch it up but instead provide fresh wasp file just for this test. But ok, we can do this in the future.

shayneczyzewski replied:
Good point, yes let me separate it out into two functions (both they have to explicitly call from a golden test) and note the fragility here. In general, I tried to keep the patches in code vs using filesystem stuff (a la fixtures/overlays). The fragility in the fresh Wasp file is if any syntax in there changes, or we add/remove stuff to default main.wasp, it too could "break" (but possibly silently, which may be worse?). I think for now with 2 separate commands we can prob live with the assumption if we put that postgres snippet in line 2 under the `app` decl, we are prob ok for now. But agree, not ideal under any scenario. :/

Martinsos replied:
Yeah, that sounds good enough for now!

Martinsos commented on waspc/e2e-test/Tests/WaspBuildTest.hs:
```diff
@@ -0,0 +1,15 @@
+module Tests.WaspBuildTest (waspBuild) where
```
Btw you don't need to suffix files with Test if you don't want to, we are doing that in unit tests because then I can give regex to Tasty which it uses to pick up tests, but you don't need that here. If you like it thought, I am ok with it, just mentioning.

Martinsos commented on waspc/e2e-test/Tests/WaspCompileTest.hs:
```diff
@@ -0,0 +1,20 @@
+            waspCliCompile
+          ]
+
+  runGoldenTest "waspCompile" makeShellCommand
```
A bit interesting that something named `GoldenTest` is calling `runGoldenTest` inside it. Is it really a golden test then? If it is running a golden test? What is it then? Not sure hm. Maybe then it is just `IO TestTree`? Did something change, did we name something else `GoldenTest` before? Or was it just `IO TestTree` all the time?
Maybe we are not running any golden test, but we are running golden test that is based on shell commands? So `runShellCmdGoldenTest`?
What is a `GoldenTest` for us really? How can it be `IO TestTree`, what does that mean?
What if `Main.hs` would have `runGoldenTest` mapped on all these tests? Or maybe those names are assigned there instead of here -> you could name the tests there!
A bunch of questions / ideas.

Maybe it is fine, maybe I am just overthinking this!

shayneczyzewski replied:
We did have it taking a param I think, but now it is just an `IO TestTree` again, so let's go back to that and not hide it. Good call 👍🏻 I can also see what mapping run over the tests looks like too.

Martinsos reviewed (commented):
Wohooo it got out of the draft :D!

A thought: maybe split this into 2 PRs, one with the code changes you did, and another one with the golden outputs? That will make this PR easier to review, right now I am digging through files to figure out which ones are not generated.

Regarding the name of the `goldens` dir -> it is fine, I like it :D, buuuuuut, what if it was named `test-outputs`? What that maybe be a better name? Then you would have `test-outputs/foo-golden` and `test-outputs/foo-current`. If you don't think this is an improvement fine, but it seems to me it would complete the naming concept of "golden and current test outputs".

shayneczyzewski commented:
> Wohooo it got out of the draft :D!
> 
> A thought: maybe split this into 2 PRs, one with the code changes you did, and another one with the golden outputs? That will make this PR easier to review, right now I am digging through files to figure out which ones are not generated.
> 
> Regarding the name of the `goldens` dir -> it is fine, I like it :D, buuuuuut, what if it was named `test-outputs`? What that maybe be a better name? Then you would have `test-outputs/foo-golden` and `test-outputs/foo-current`. If you don't think this is an improvement fine, but it seems to me it would complete the naming concept of "golden and current test outputs".

Thanks for the review! Good point on splitting into 2 PRs, I will remove all golden outputs and then start on the feedback! Appreciate it 👍🏻

shayneczyzewski commented on waspc/e2e-test/GoldenTest.hs:
```diff
@@ -0,0 +1,77 @@
+          let goldenOutputAbsFp = unpack $ replace (pack $ currentOutputDirAbsFp context) (pack $ goldenOutputDirAbsFp context) (pack currentOutputAbsFp)
+      ]
+  where
+    -- TODO: Ideally we would not ignore `package.json`, but in CI the order of dependencies differs. :/
```
It unfortunately is `package.json`. It seems for whatever reason we were writing them in a different order locally vs on the CI server. I'm not 100% sure atm if it was because of any NPM changes on `main` I didn't merge in locally, but I will test and revisit, and even create an Issue as @faassen suggested before if it still breaks. I forgot we generated that file ourselves so writing deterministically would be nice.

Martinsos replied:
That sounds weird! It should be deterministic, even without sorting, since order is currently defined by how dependencies are listed in our code.

It doesn't make much sense that CI orders stuff differently, either. There is nothing different happening there.

We should try to figure out, as maybe it is hiding some bigger problem. And having package.json in golden tests is actually pretty valuable.

Martinsos replied:
Keep in mind that recently Martijn was changing this piece of code, so maybe that changed how package.json was generated (the order), and you somehow stumbled right onto it -> version before his changes and version after them?

shayneczyzewski commented on waspc/e2e-test/GoldenTest.hs:
```diff
@@ -0,0 +1,77 @@
+          currentOutputAbsFp
+          (return ()) -- A no-op command that normally generates the file under test, but we did that in bulk above.
+        | currentOutputAbsFp <- manifestAbsFp : currentOutputAbsFps,
+          let goldenOutputAbsFp = unpack $ replace (pack $ currentOutputDirAbsFp context) (pack $ goldenOutputDirAbsFp context) (pack currentOutputAbsFp)
```
That is a good point a helper could make it easier to see.. something like:

```haskell
  let remapCurrentPathToGolden fp = unpack $ replace (pack $ currentOutputDirAbsFp context) (pack $ goldenOutputDirAbsFp context) (pack fp)

  return $
    ...
          (return ()) -- A no-op command that normally generates the file under test, but we did that in bulk above.
        | currentOutputAbsFp <- manifestAbsFp : currentOutputAbsFps,
          let goldenOutputAbsFp = remapCurrentPathToGolden currentOutputAbsFp
      ]
```

Martinsos replied:
That also helps! I thought you could extract the whole line, like this:

```hs
  let goldenOutputAbsFp = unpack $ replace (pack $ currentOutputDirAbsFp context) (pack $ goldenOutputDirAbsFp context) (pack currentOutputAbsFp)

  return $
    ...
          (return ()) -- A no-op command that normally generates the file under test, but we did that in bulk above.
        | currentOutputAbsFp <- manifestAbsFp : currentOutputAbsFps
      ]
```

shayneczyzewski replied:
I don't think we can do that since we need `currentOutputAbsFp` from inside the list comprehension, so I would need to pass that in somehow. The other two are available outside, however.

Martinsos replied:
Ouf sorry completely missed that! Ok, makes sense!

shayneczyzewski commented on waspc/e2e-test/GoldenTest.hs:
```diff
@@ -0,0 +1,77 @@
+    -- TODO: Ideally we would not ignore `package.json`, but in CI the order of dependencies differs. :/
+    dirFilter fp =
+      return $ notElem (takeFileName fp) [".waspinfo", ".gitignore", "node_modules", "dev.db", "dev.db-journal", "package.json", "package-lock.json", "golden.manifest"]
+    getAllDirFilesRecursivelyFiltered dirToFilterAbsFp =
```
Good points on `list` vs `get` and filtered confusion. I like the `listRelevantTestOutputFiles` suggestion 👍🏻 thanks!

shayneczyzewski commented on waspc/e2e-test/GoldenTest.hs:
```diff
@@ -0,0 +1,77 @@
+      ]
+  where
+    -- TODO: Ideally we would not ignore `package.json`, but in CI the order of dependencies differs. :/
+    dirFilter fp =
```
Yep, you are correct. Terribly opaque original name for filter 😬 Name and position both good points 👍🏻

shayneczyzewski commented on waspc/e2e-test/GoldenTest.hs:
```diff
@@ -0,0 +1,77 @@
+      return $ notElem (takeFileName fp) [".waspinfo", ".gitignore", "node_modules", "dev.db", "dev.db-journal", "package.json", "package-lock.json", "golden.manifest"]
+    getAllDirFilesRecursivelyFiltered dirToFilterAbsFp =
+      getDirFiltered dirFilter dirToFilterAbsFp >>= filterM doesFileExist
+    writeFileManifest baseAbsFp filePaths manifestAbsFp = do
```
Good point, they are getting complex enough to warrant type sigs 👍🏻

shayneczyzewski commented on waspc/e2e-test/ShellCommands.hs:
```diff
@@ -0,0 +1,83 @@
+
+-- Each shell command gets access to the current project name and current/golden output directories.
+data ShellCommandContext = ShellCommandContext
+  { currentProjectName :: String,
+    currentOutputDirAbsFp :: FilePath,
+    goldenOutputDirAbsFp :: FilePath
```
Yep, can do, appreciate the convention note! 👍🏻

shayneczyzewski replied:
I will also use an underscore, since that seems to be our convention in the code base, so something like: `_ctxtCurrentProjectName`...

shayneczyzewski commented on waspc/e2e-test/ShellCommands.hs:
```diff
@@ -0,0 +1,83 @@
+combineShellCommands = intercalate " && "
+
+combineMakeShellCommands :: [MakeShellCommand] -> MakeShellCommand
+combineMakeShellCommands commands context = combineShellCommands (map (\command -> command context) commands)
```
My head told me that feels verbose, but did not see `$` coming 🤯 But makes sense 🚀 Super cool

Martinsos commented on waspc/e2e-test/ShellCommands.hs:
```diff
@@ -0,0 +1,83 @@
+type ShellCommand = String
+
+-- A way to create a shell command given some context info.
+type MakeShellCommand = ShellCommandContext -> ShellCommand
```
Ha I thought you will like the idea of using Reader :D! I forgot about your suggestion for `cwd`, but now I think I do remember. Ok makes sense, let's see how that turns out -> it should be nicer as you will not have to thread context everywhere anymore, and I saw that on many places you ignore it, so makes sense to do it this way.

shayneczyzewski replied:
This turned out to be pretty cool! :D Thanks for nudging me to do it now, I think it feels a lot cleaner and more coherent.

1) So we now have a `ShellCommandBuilder` that is a `Reader` over `ShellCommandContext`, which all our shell command helper functions are in now:
```haskell
newtype ShellCommandBuilder a = ShellCommandBuilder {_runShellCommandBuilder :: Reader ShellCommandContext a}
  deriving (Functor, Applicative, Monad, MonadReader ShellCommandContext)

runShellCommandBuilder :: ShellCommandBuilder a -> ShellCommandContext -> a
runShellCommandBuilder shellCommandBuilder context =
  runReader (_runShellCommandBuilder shellCommandBuilder) context

...

cdIntoCurrentProject :: ShellCommandBuilder ShellCommand
cdIntoCurrentProject = do
  context <- ask
  return $ "cd " ++ _ctxtCurrentProjectName context
```

2) I re-introduced a `GoldenTest`, but this time makes more sense. :) It has a test name and a series of commands (within the `ShellCommandBuilder` context). Now the type sig of `runGoldenTest` also feels right:
```haskell
data GoldenTest = GoldenTest {_goldenTestName :: String, _goldenTestCommands :: ShellCommandBuilder [ShellCommand]}

makeGoldenTest :: String -> ShellCommandBuilder [ShellCommand] -> GoldenTest
makeGoldenTest name commands = GoldenTest {_goldenTestName = name, _goldenTestCommands = commands}

-- | This runs a golden test by creating a Wasp project (via `wasp-cli new`), running commands,
-- and then comparing all file outputs to the corresponding golden test output directory.
runGoldenTest :: GoldenTest -> IO TestTree
runGoldenTest goldenTest = do
  ...

  let shellCommand = combineShellCommands $ runShellCommandBuilder (_goldenTestCommands goldenTest) context
  putStrLn $ "Running the following command: " ++ shellCommand

  ...
```

3) Each test is now a `GoldenTest` and we build the commands pretty simply:
```haskell
waspCompile :: GoldenTest
waspCompile = do
  let commands =
        sequence
          [ waspCliNew,
            cdIntoCurrentProject,
            waspCliCompile
          ]

  makeGoldenTest "waspCompile" commands
```

4) Now in `Main.hs` we `mapM runGoldenTest` over all our `GoldenTest`s:
```haskell
tests :: IO TestTree
tests = do
  testGroup "All Golden Dir Tests"
    <$> mapM runGoldenTest [waspNew, waspCompile, waspMigrate, waspBuild]
```

Martinsos commented on waspc/e2e-test/ShellCommands.hs:
```diff
@@ -0,0 +1,83 @@
+
+appendToWaspFile :: String -> MakeShellCommand
+appendToWaspFile content _context =
+  -- NOTE: Using `show` to preserve newlines in string.
```
Btw certainly don't bother with it now, I think that would be too much.

matijaSos commented on waspc/e2e-test/ShellCommands.hs:
```diff
@@ -0,0 +1,97 @@
+
+waspCliNew :: ShellCommandBuilder ShellCommand
+waspCliNew = do
+  context <- ask
```
nice - Reader in action!!

Martinsos replied:
Cool indeed :)!

Martinsos commented on waspc/e2e-test/ShellCommands.hs:
```diff
@@ -0,0 +1,114 @@
+
+-- Each shell command gets access to the current project name, and maybe other things in future.
+data ShellCommandContext = ShellCommandContext
+  { _ctxtCurrentProjectName :: String
```
Interesting, didn't you have more stuff in the context a bit before?

shayneczyzewski replied:
Yeah, I had the paths to the current and golden dirs in there, but turns out they were not used in shell commands (yet) so I removed them. But I think this place will grow over time if we add more complex shell commands 👍🏻

Martinsos commented on waspc/e2e-test/Tests/WaspBuildTest.hs:
```diff
@@ -0,0 +1,23 @@
+            reformatPackageJson
+          ]
+
+  makeGoldenTest "waspBuild" commands
```
Just curious, although it makes no big difference: why not merge it like this?

```hs
  makeGoldenTest "waspBuild" $ sequence
          [ waspCliNew,
            cdIntoCurrentProject,
            setDbToPSQL,
            waspCliBuild,
            reformatPackageJson
          ]
```

shayneczyzewski replied:
Yeah that works, can update! 👍🏻

Martinsos commented on waspc/e2e-test/Tests/WaspNewTest.hs:
```diff
@@ -0,0 +1,9 @@
+  )
+
+waspNew :: GoldenTest
+waspNew = makeGoldenTest "waspNew" ((: []) <$> waspCliNew)
```
Woah woah doing some magic here :D :P.
Btw, just in case you didn't know, you could also have done
```hs
waspNew = makeGoldenTest "waspNew" (singleton <$> waspCliNew)
```
or
```hs
waspNew = makeGoldenTest "waspNew" (pure <$> waspCliNew)
```

shayneczyzewski replied:
Arg yeah lol `pure` muuuch easier haha thnx

Martinsos reviewed (approved):
Super nice @shayneczyzewski , that Reader is looking nice!

Left a couple very minor comments, but all is good otherwise. I approved! Awesome contribution to Wasp!!

shayneczyzewski commented:
> Super nice @shayneczyzewski , that Reader is looking nice!
> 
> Left a couple very minor comments, but all is good otherwise. I approved! Awesome contribution to Wasp!!

Thanks @Martinsos I appreciate it! :D This is probably our PR with the most collective comments yet, but I think it was helpful to go from a 1-day prototype to something we are all comfortable working with for years to come! 🚀 

The only thing you may have missed is the actual process of reformatting the `package.json` file. In short, we have it defined like so (in `ShellCommands.hs`):
```haskell
data OutputDir = DevOutputDir | BuildOutputDir

-- Uses python2's json.tool to sort keys as `npm install` overwrites
-- package.json when writing package-lock.json and different versions produce different ouput.
-- NOTE: python3 won't sort them, but we want to, so force python2.
-- Ref: https://github.com/wasp-lang/wasp/issues/482
reformatPackageJson :: OutputDir -> ShellCommandBuilder ShellCommand
reformatPackageJson outputDir =
  return $ combineShellCommands $ concatMap (reformatJsonCommands . componentPackageJsonFp) ["server", "web-app"]
  where
    componentPackageJsonFp :: String -> FilePath
    componentPackageJsonFp component = case outputDir of
      DevOutputDir -> ".wasp/out/" ++ component ++ "/package.json"
      BuildOutputDir -> ".wasp/build/" ++ component ++ "/package.json"

    reformatJsonCommands :: FilePath -> [ShellCommand]
    reformatJsonCommands fp =
      let tmpFp = fp ++ ".tmp"
       in [ "python2 -m json.tool " ++ fp ++ " > " ++ tmpFp,
            "mv " ++ tmpFp ++ " " ++ fp
          ]
```

And use it in the test files like so: `reformatPackageJson DevOutputDir`. I wanted to keep it an explicit part of our commands so we remember we are doing it, and the param supports the standard `out` dir and `build` dir, depending on usage.

If that looks sane, I can go ahead and merge Part 1, and then retarget Part 2 for `main` and after any last minute feedback get the golden test dirs actually in there. Thanks!

Martinsos reviewed (approved):
Thanks for sharing the details @shayneczyzewski -> looks good to me!

faassen commented on waspc/e2e-test/GoldenTest.hs:
```diff
@@ -0,0 +1,85 @@
+  let goldenOutputDirAbsFp = testOutputsDirAbsFp FP.</> (_goldenTestName goldenTest ++ "-golden")
+
+  -- Remove existing current output files from a prior test run.
+  callCommand $ "rm -rf " ++ currentOutputDirAbsFp
```
I hope this can never get a handle to my home directory...

shayneczyzewski replied:
Strong path and prior checks for `waspc` have your back ;) We unfortunately have to remove the `current` dirs somehow...

faassen commented on waspc/e2e-test/ShellCommands.hs:
```diff
@@ -0,0 +1,121 @@
+
+data OutputDir = DevOutputDir | BuildOutputDir
+
+-- Uses python2's json.tool to sort keys as `npm install` overwrites
```
This is a pretty flaky dependency - python 2 is not being maintained anymore, may not be on all platforms, and we're depending on a whole different programming language. It works for now, but this should really be replaced with something more reliable in the future - I wonder whether Aeson has an ordering concept for instance.

faassen replied:
It appears this https://hackage.haskell.org/package/aeson-pretty-0.8.9/docs/Data-Aeson-Encode-Pretty.html
has a way to sort keys (confCompare) that can be defined.

shayneczyzewski replied:
There are sadly a few fragile things in here we can improve, I'll add a more explicit TODO around that part 👍🏻
